go/vt/vttablet/grpctmclient/cached_client_test.go

```diff
@@ -0,0 +1,284 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpctmclient
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"net"
+	"os"
+	"runtime"
+	"runtime/pprof"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/net/nettest"
+	"google.golang.org/grpc"
+
+	"vitess.io/vitess/go/sync2"
+	"vitess.io/vitess/go/vt/vttablet/grpctmserver"
+	"vitess.io/vitess/go/vt/vttablet/tabletmanager"
+	"vitess.io/vitess/go/vt/vttablet/tmrpctest"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+)
+
+func grpcTestServer(t testing.TB, tm tabletmanager.RPCTM) (*net.TCPAddr, func()) {
+	t.Helper()
+
+	lis, err := nettest.NewLocalListener("tcp")
+	if err != nil {
+		t.Fatalf("Cannot listen: %v", err)
+	}
+
+	s := grpc.NewServer()
+	grpctmserver.RegisterForTest(s, tm)
+	go s.Serve(lis)
+
+	var shutdownOnce sync.Once
+
+	return lis.Addr().(*net.TCPAddr), func() {
+		shutdownOnce.Do(func() {
+			s.Stop()
+			lis.Close()
+		})
+	}
+}
+
+func BenchmarkCachedConnClient(b *testing.B) {
+	tmserv := tmrpctest.NewFakeRPCTM(b)
+	tablets := make([]*topodatapb.Tablet, 4)
+	for i := 0; i < len(tablets); i++ {
+		addr, shutdown := grpcTestServer(b, tmserv)
+		defer shutdown()
+
+		tablets[i] = &topodatapb.Tablet{
+			Alias: &topodatapb.TabletAlias{
+				Cell: "test",
+				Uid:  uint32(addr.Port),
+			},
+			Hostname: addr.IP.String(),
+			PortMap: map[string]int32{
+				"grpc": int32(addr.Port),
+			},
+		}
+	}
+
+	b.ResetTimer()
+	client := NewCachedConnClient(5)
+	defer client.Close()
+
+	for i := 0; i < b.N; i++ {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*100)
+		x := rand.Intn(len(tablets))
+		err := client.Ping(ctx, tablets[x])
+		if err != nil {
+			b.Errorf("error pinging tablet %v: %w", tablets[x].Hostname, err)
+		}
+
+		cancel()
+	}
+}
+
+func TestCachedConnClient(t *testing.T) {
+	t.Parallel()
+
+	if os.Getenv("VT_PPROF_TEST") != "" {
+		file, err := os.Create(fmt.Sprintf("%s.profile.out", t.Name()))
+		require.NoError(t, err)
+		defer file.Close()
+		if err := pprof.StartCPUProfile(file); err != nil {
+			t.Errorf("failed to start cpu profile: %v", err)
+			return
+		}
+		defer pprof.StopCPUProfile()
+	}
```

Comment:
By the way, you can get this exact same behavior by simply passing the `-cpuprofile` flag to `go test` :+1: